### PR TITLE
Weather UI using RecyclerView and CardView

### PR DESCRIPTION
**Pull request by brenpardal**
### **WHAT'S NEW**
**This update includes:**
* Development of the WeatherApp UI
* Show the weather information for the next five days into a cardView in a RecyclerView

#### *Preview Device Nexus S:*
nexus_s)

#### *Preview DevicePixel 2XL:*
pixel_2xl)

#### *Preview no internet connection:*
no_internet_connection)

### CODE REVIEW

File: app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(weather[position])
+    }
+
+    override fun getItemCount() = weather.size
+
+    companion object {
+        private const val URL = "http://openweathermap.org/img/wn/"
+        private const val FORMAT = "@2x.png"
+        private const val CELSIUS = "°C"
```
Comment (SartiMau) [inline app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt:42]: put this string in the resources pls

File: app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        private const val REAL_FEEL = "RealFeel "
```
Comment (SartiMau) [inline app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt:43]: same

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt (hunk 1 above)
Comment (brenpardal) [reply app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt:42]: Solved!

File: app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt (hunk 2 above)
Comment (brenpardal) [reply app/src/main/java/com/weatherapp/adapter/WeatherAdapter.kt:43]: Solved!

File: app/src/main/java/com/weatherapp/data/service/WeatherService.kt [hunk 3]
```diff
@@ -11,12 +11,12 @@ class WeatherService {
     private val api = WeatherRequestGenerator()
     private val mapper = WeatherMapper()
 
-    fun getWeather(id: String): Observable<Weather> {
+    fun getWeather(id: String): Observable<List<Data>> {
```
Comment (nicolbidone) [inline app/src/main/java/com/weatherapp/data/service/WeatherService.kt:14]: please give Data class a more informative name like WeatherData

File: app/src/main/java/com/weatherapp/data/service/WeatherService.kt [hunk 4]
```diff
@@ -11,12 +11,12 @@ class WeatherService {
...
         return Observable.create { subscriber ->
             val callResponse = api.createService(WeatherApi::class.java).getCurrentWeatherData(id, BuildConfig.WEATHER_API_KEY, UNITS)
             val response = callResponse.execute()
             if (response.isSuccessful) {
-                response.body()?.let { subscriber.onNext(mapper.transform(it)) }
+                response.body()?.list?.filter { it.date.contains(HOUR) }?.let { subscriber.onNext(mapper.transformData(it)) }
```
Comment (nicolbidone) [inline app/src/main/java/com/weatherapp/data/service/WeatherService.kt:19]: please give transformData a more informative name like transformWeatherData

File: app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt [hunk 5]
```diff
@@ -2,30 +2,31 @@ package com.weatherapp.data.mapper
...
+            auxData.add(
+                Data(
+                    formatApp.format(formatJson.parse(it.date)),
+                    transformMain(it.main),
+                    transformWeatherDescription(it.weatherDescription)
+                )
+            )
         }
         return auxData
     }
```
Comment (nicolbidone) [inline app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt:27]: here you can use " = this.map { it.mapToWeatherData() } " and use an extension function to map the data

File: app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt [hunk 6]
```diff
@@ -2,30 +2,31 @@ package com.weatherapp.data.mapper
...
 import com.weatherapp.data.entity.Temperature
-import com.weatherapp.data.entity.Weather
 import com.weatherapp.data.entity.WeatherDescription
 import com.weatherapp.data.service.response.DataResponse
 import com.weatherapp.data.service.response.MainResponse
 import com.weatherapp.data.service.response.WeatherDescriptionResponse
-import com.weatherapp.data.service.response.WeatherResponse
+import java.text.SimpleDateFormat
 
 class WeatherMapper {
```
Comment (nicolbidone) [inline app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt:11]: why do you wrap the mappers in a class?

File: app/src/main/res/layout/activity_main.xml [hunk 7]
```diff
@@ -4,4 +4,25 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:drawableRight="@mipmap/ic_launcher"
+        android:text="@string/app_name"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/recycler_view_weather"
+        android:layout_width="0dp"
```
[the remark this answers is not in the document]
Comment (brenpardal) [reply app/src/main/res/layout/activity_main.xml:21]: Thanks. Solved!

File: app/src/main/java/com/weatherapp/data/service/WeatherService.kt (hunk 3 above)
Comment (brenpardal) [reply app/src/main/java/com/weatherapp/data/service/WeatherService.kt:14]: Solved.

File: app/src/main/java/com/weatherapp/data/service/WeatherService.kt (hunk 4 above)
Comment (brenpardal) [reply app/src/main/java/com/weatherapp/data/service/WeatherService.kt:19]: Solved.

File: app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt (hunk 5 above)
Comment (brenpardal) [reply app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt:27]: Solved. I used an extension function to map the data.

File: app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt (hunk 6 above)
Comment (brenpardal) [reply app/src/main/java/com/weatherapp/data/mapper/WeatherMapper.kt:11]: Solved.  I use an extension function to map the data now.

File: app/src/main/res/layout/weather_item.xml [hunk 8]
```diff
@@ -0,0 +1,76 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.cardview.widget.CardView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    style="@style/BackgroundItemStyle"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    app:layout_constraintBottom_toBottomOf="parent"
+    app:layout_constraintEnd_toEndOf="parent"
+    app:layout_constraintStart_toStartOf="parent"
+    app:layout_constraintTop_toTopOf="parent">
```
Comment (mauroLonghi) [inline app/src/main/res/layout/weather_item.xml:10]: 1)Why you are using constraints guides if dont have a constriant as a parent?
2) why use a wrap size iff you need to show in all the screen?

**Review by SartiMau — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

File: app/src/main/res/layout/weather_item.xml (hunk 8 above)
Comment (brenpardal) [reply app/src/main/res/layout/weather_item.xml:10]: Solved.
1) Removed this. 
2) Replaced by match_parent.

**Review by mauroLonghi — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Pull request merged**